### PR TITLE
refactor: refactor client backend

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #3 
fixes: #5 
fixes: #6

Refactor.
I am about to add a new `ExplorerAPI` method - for getting transactions for an account - so getting this ready to be worked on, make the next PR easier.

*After This*

1. I need to add support for the API key (to prevent getting rate limited when fetching transaction lists back-to-back).
2. Wait for ape release
3. Open PR for new explorer API method impl. `get_account_transactions()`. I have the work ready just relies on this + ape release.

### How I did it

Refactored!

### How to verify it

Things still work

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: tests/test_etherscan.py [hunk 1]
```diff
@@ -1,21 +1,63 @@
...
+        # Request will fail if made with incorrect parameters.
+        assert base_uri == "https://api.etherscan.io/api"
+        assert params == expected_params
+        return etherscan_abi_response
+
+    get_patch.get.side_effect = get
+    return get_patch
+
+
+def test_get_address_url(explorer):
```
Comment (fubuloubu) [inline tests/test_etherscan.py:39]: Can we turn these into `@pytest.mark.parametrize` cases (with one entry to start) so it's easy to add new cases to try for?

Also when this plugin supports explorers for multiple networks (in fact should probably try it with all the eth testnets)

This and the transaction one and contract type one

File: tests/test_etherscan.py [hunk 2]
```diff
@@ -1,21 +1,63 @@
 import json
+from pathlib import Path
 
 import pytest
 from ape import networks
 from requests import Response
 
+ADDRESS = "0xab5801a7d398351b8be11c439e05c5b3259aec9b"
+TRANSACITON = "0x0da22730986e96aaaf5cedd5082fea9fd82269e41b0ee020d966aa9de491d2e6"
```
Comment (fubuloubu) [inline tests/test_etherscan.py:9]: spelt wrong

File: tests/test_etherscan.py (hunk 1 above)
Comment (fubuloubu) [reply tests/test_etherscan.py:39]: also maybe try different types contracts at different addresses (proxies, no contract at address, unpublished, etc.)

Comment (unparalleled-js) [reply tests/test_etherscan.py:39]: I can definitely add tests for the test nets.
I don't really know how different contract types with affect the outcome on our side... we just return a str address?

re: the `parametrize()`, I am not exactly sure what you mean! What is being parametrized here?

Comment (fubuloubu) [reply tests/test_etherscan.py:39]: Oh whoops, I mean for get contract type

parametrize the networks (and addresses, if relevant)

Comment (unparalleled-js) [reply tests/test_etherscan.py:39]: Oh, would the ABI serialization stuff differ in this case? Re: the different types of contracts

Comment (fubuloubu) [reply tests/test_etherscan.py:39]: Yes, it might!

Comment (unparalleled-js) [reply tests/test_etherscan.py]: Okay I parametrized the network, the expected URL, and the address / txn has (testing one of the testnets too).

I will admit I am not a huge fan of parametrizing tests to this level, but that is just my personal preference.

Testing the contract types stuff will take some more research on my end.

File: tests/test_etherscan.py [hunk 3]
```diff
@@ -36,27 +39,38 @@ def get(base_uri, params=None, *args, **kwargs):
...
-    actual = explorer.get_address_url(ADDRESS)
+@pytest.mark.parametrize(
+    "network,expected_prefix,address",
+    [(NETWORKS[0], "etherscan.io", ADDRESS), (NETWORKS[1], "ropsten.etherscan.io", ADDRESS)],
+)
+def test_get_address_url(network, expected_prefix, address):
+    expected = f"https://{expected_prefix}/address/{ADDRESS}"
+    explorer = get_explorer(network)
+    actual = explorer.get_address_url(address)  # type: ignore
     assert actual == expected
```
Comment (fubuloubu) [inline tests/test_etherscan.py:49]: hmm, maybe a better test would be to see if it gives you a 404

Comment (unparalleled-js) [reply tests/test_etherscan.py:49]: Hmm, I disagree. Making real network requests in tests is not a good thing.

Comment (unparalleled-js) [reply tests/test_etherscan.py:49]: Unless you control the service. Making requests to Etherscan / Infura, etc. in tests, even integration, is not really necessary and just complicates stuff. If Etherscan or Infura break, that is not on us. That should not cause test failures in `ape`. Relying on services in tests can lead to other behaviors too. It is out of scope!

What would be _in scope_ though, is testing how we handle 404s or something (we are not doing that here but `ape` maybe we would?)

Comment (fubuloubu) [reply tests/test_etherscan.py:49]: fair

File: tests/test_etherscan.py [hunk 4]
```diff
@@ -1,21 +1,79 @@
...
+        "address": ADDRESS,
+    }
+    setup_mock_get(mocker, etherscan_abi_response, expected_params)
+
+    explorer = get_explorer("mainnet")
+    actual = explorer.get_contract_type(ADDRESS)  # type: ignore
+
+    # Name comes from the 'get_contract_response.json' file
+    expected = "BoredApeYachtClub"
+    assert actual.contractName == expected
```
Comment (fubuloubu) [inline tests/test_etherscan.py:79]: This is the one that would work well with parameterization (maybe with integration test)

Try some contracts and check if you get a contract type name or nothing

But maybe this is all overkill, as you said in the other comments re: rate limiting and directly testing etherscan

Comment (unparalleled-js) [reply tests/test_etherscan.py:79]: We are actually mocking the response here in the existing test.

I would like to make sure we handle these edge cases though. If the responses differ, we can still add tests making sure we handle those situations.

Do you have any good examples of proxy contracts e.g. stuff you like me to test here?

Comment (fubuloubu) [reply tests/test_etherscan.py:79]: example proxy: https://github.com/yearn/yearn-vaults/blob/main/contracts/Registry.vy#L164

This contract (`Registry`) collects new deployments of the vaults, and allows you to make EIP-1167 proxies to them.

Here's one in prod:
- `Registry`: https://etherscan.io/address/v2.registry.ychad.eth#code
- `Vault`: https://etherscan.io/address/0x9C13e225AE007731caA49Fd17A41379ab1a489F4#code
- EIP-1167 proxy to `Vault`: https://etherscan.io/address/0xa354F35829Ae975e850e23e9615b11Da1B3dC4DE#code

**Review by fubuloubu — APPROVED** [review]
